### PR TITLE
P3A: add support for multihash attestation fingerprints

### PR DESCRIPTION
**Pull request by rillian**
The current scheme for encoding the TLS certificate fingerprint in the AWS Nitro secure enclave attestation document is [hard to parse](https://github.com/brave/nitriding-daemon/issues/23) and we want to switch it to a sequence of multihash values, which are self-delimiting and more compact.

As the first step of this transition, support both the old and new schema, determined by the field length, which is different in the two encodings.

Resolves https://github.com/brave/brave-browser/issues/31131

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

Unit test coverage should be sufficient for this until we transition the production service. I'm marked the corresponding issue QA/No.

### CODE REVIEW

File: components/p3a/nitro_utils/attestation.cc [hunk 1]
```diff
@@ -35,16 +36,28 @@ namespace nitro_utils {
...
 namespace {
 
-constexpr char kHashPrefix[] = "sha256:";
-constexpr size_t kHashPrefixLength = 7;
-constexpr size_t kSHA256HashLength = 32;
-constexpr size_t kUserDataMinLength = kSHA256HashLength + kHashPrefixLength;
 constexpr size_t kAttestationBodyMaxSize = 16384;
+// The user_data field contains a multihash of the TLS cert
+constexpr size_t kMultihashPrefixLength = 2;
+constexpr size_t kMultihashSHA256Code = 0x12;
```
Comment (fmarier) [inline components/p3a/nitro_utils/attestation.cc:42]: It might be nice to add https://github.com/multiformats/multicodec/blob/b98f2f38fc63c992545263815c2afdd692d2dc8e/table.csv#L9 as a comment here

File: components/p3a/nitro_utils/attestation.cc [hunk 2]
```diff
@@ -35,16 +36,28 @@ namespace nitro_utils {
 
 namespace {
 
-constexpr char kHashPrefix[] = "sha256:";
-constexpr size_t kHashPrefixLength = 7;
-constexpr size_t kSHA256HashLength = 32;
-constexpr size_t kUserDataMinLength = kSHA256HashLength + kHashPrefixLength;
 constexpr size_t kAttestationBodyMaxSize = 16384;
+// The user_data field contains a multihash of the TLS cert
```
Comment (fmarier) [inline components/p3a/nitro_utils/attestation.cc:40]: We could link to https://multiformats.io/multihash/#the-multihash-format here for people unfamiliar with that format.

File: components/p3a/nitro_utils/attestation.cc (hunk 1 above)
Comment (rillian) [reply components/p3a/nitro_utils/attestation.cc:44]: Addressed in 9b1cfccaca47fded6486b47f3fb306764d9e284c

File: components/p3a/nitro_utils/attestation.cc (hunk 2 above)
Comment (rillian) [reply components/p3a/nitro_utils/attestation.cc]: Good idea, thanks! Added in 9b1cfccaca47fded6486b47f3fb306764d9e284c.

File: components/p3a/nitro_utils/attestation.h [hunk 3]
```diff
@@ -26,6 +28,20 @@ void RequestAndVerifyAttestationDocument(
     base::OnceCallback<void(scoped_refptr<net::X509Certificate>)>
         result_callback);
 
+// Functions used for unit tests
+//
+// These take the cbor serialization of just the attestation document
+// portion of the COSE Sign1 object returned by the AWS Nitro enclave
+// in reponse to remote attestation requests.
```
Comment (DJAndries) [inline components/p3a/nitro_utils/attestation.h:35]: nit: typo "response"

File: components/p3a/nitro_utils/attestation.cc [hunk 4]
```diff
@@ -266,4 +303,25 @@ void RequestAndVerifyAttestationDocument(
...
+    LOG(ERROR) << "Nitro verification: expected cbor map for test";
+    return false;
+  }
+  return VerifyNonce(document.value().GetMap(), orig_nonce);
+}
+
+bool VerifyUserDataKeyForTesting(
+    const std::vector<uint8_t>& attestation_bytes,
+    scoped_refptr<net::X509Certificate> expected_cert) {
+  auto document = cbor::Reader::Read(attestation_bytes);
```
Comment (DJAndries) [inline components/p3a/nitro_utils/attestation.cc:319]: it seems that you could just have a helper method in `NitroAttestationTest` that handles the parsing of the cbor, and then you could pass the result directly to `VerifyUserDataKey` and `VerifyNonce`

**Review by DJAndries — CHANGES_REQUESTED** [review]
looks great, thanks for adding the test. only a few small comments

**Review by fmarier — APPROVED** [review]

File: components/p3a/nitro_utils/attestation.cc (hunk 4 above)
Comment (rillian) [reply components/p3a/nitro_utils/attestation.cc:319]: I was trying to follow the style guide about [testing-specific code](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md#tests-and-test_only-code) to maintain [internal linkage](https://google.github.io/styleguide/cppguide.html#Internal_Linkage) for the actual verification functions.

Seems a bit clunky either way. You prefer exposing them?

Comment (DJAndries) [reply components/p3a/nitro_utils/attestation.cc:319]: ah sorry, for some reason I thought that the two functions were already exposed. I think this looks good as is. Maybe it would be good to move the cbor reading code to a function in the anonymous namespace to make the code more DRY, but up to you.

Comment (rillian) [reply components/p3a/nitro_utils/attestation.cc:319]: Ok, thanks for clarifying.

It's only a couple of lines, and we'd still have to check the return code, so I'll leave the duplication. If there were three copies, it would be worth a separate function.

**Comment by rillian** [comment]
Squashed review fixes.

**Review by DJAndries — APPROVED** [review]
lgtm, thanks

File: components/p3a/nitro_utils/attestation.cc [hunk 5]
```diff
@@ -89,30 +104,52 @@ bool VerifyUserDataKey(scoped_refptr<net::X509Certificate> server_cert,
...
-        << "Nitro verification: server cert fp does not match user data fp, "
-        << "user data = "
-        << base::HexEncode(user_data_it->second.GetBytestring())
-        << ", server cert fp = " << base::HexEncode(server_cert_fp.data);
-    return false;
+
+  // The hashes in the  old and new user_data schemes have incommensurate
+  // lengths, so use the total length to distinguish between them.
+  if (user_data_bytes.size() == kUserDataOldLength) {
+    if (memcmp(user_data_bytes.data(), kHashPrefix, kHashPrefixLength) != 0) {
```
[the remark this answers is not in the document]
Comment (thypon) [reply components/p3a/nitro_utils/attestation.cc:120]: Unless this is executed in a secure-element/separated context (a local/remote environment), the user can already read the data.
Are any page/tab context able to access an oracle out of this comparison? If the answer is yes it is indeed a local/remote environment.

I think if we have any doubt about if this will have a oracle or not, it is worth changing to `crypto::SecureMemEqual`. The cost should be negligible.
